specs/agents/configuration.md

```diff
@@ -69,3 +69,33 @@ Failure to process one config attribute should not affect processing of others.
 #### Feature flag
 
 Agents should implement a [configuration option](https://docs.google.com/spreadsheets/d/1JJjZotapacA3FkHc2sv_0wiChILi3uKnkwLTjtBmxwU), (`CENTRAL_CONFIG`) which lets users disable the central configuration polling.
+
+
+### Configuration options
+
+This section specifies *some* configuration options common to two or more APM
+agents. Some configuration options are discussed elsewhere in this repository
+if they better match a particular section -- for example
+[`transaction_ignore_urls`](./tracing-instrumentation-http.md#transaction_ignore_urls-configuration).
+
+#### `ELASTIC_APM_DISABLE_SEND`
+
+`ELASTIC_APM_DISABLE_SEND` is a boolean configuration option to have an APM agent
+be enabled but not communicate with an APM server. Use cases for setting this
+`true` include getting the following, **without** having deployed an APM server
+for event collection.
+
+- maintaining the ability to create traces and log trace/transaction/span IDs
+  through the log correlation feature, and
+- getting automatic distributed tracing via the https://w3c.github.io/trace-context/[W3C HTTP headers]
+
+Agents that implement this configuration option:
+
+- MUST NOT attempt to communicate with APM server. This includes central configuration.
+- MUST NOT log warnings/errors related to failures to communicate with APM server.
+- MUST continue to propagate trace headers (`traceparent`, `tracestate`, etc.)
+  per normal.
+- MUST continue to support [log correlation](./log-correlation.md)
+- SHOULD attempt to reduce runtime overhead where possible. For example,
+  because events will be dropped there is no need to collect stack traces,
+  collect metrics, or to calculate breakdown metrics.
```

Comment:
I don't agree with this one. I like to think of `DISABLE_SEND` as a way to run *all* of the agent *except only* the parts that communicate with APM Server. That means doing everything as usual up until the very last point before opening a connection. If perf is an issue, users can use the `RECORDING` or `ENABLED` options instead.

Reply:
But a big use case for this option is to allow for distributed context propagation and log correlation. To reduce overhead for that use case, it does make sense to turn off features that don't have any noticeable side-effects but add overhead. If `recording` or `enabled` is set to `false`, log correlation doesn't work either.

> I like to think of DISABLE_SEND as a way to run all of the agent except only the parts that communicate with APM Server. That means doing everything as usual up until the very last point before opening a connection.

What's the advantage of not disabling side-effect-free features such as breakdown metrics if `disable_send` is `false`? Is it something you need for testing? Is it because the option name doesn't reflect that? Any suggestions for a better option name while we're standardizing on the behavior?

Reply:
This is maybe only my own use case, but for my own apps I like to run the agent as integrated as possible, doing as many things as possible, to test that nothing breaks. However, I don't need the resulting data as it's not really relevant when in development/test mode.

It may be because I am the only person who's both an agent AND application developer, but I can see other folks wanting something like this too. Like, staging or whatever, where you want to mirror prod as closely as possible but don't wan't the burden of setting up a dedicated APM Server for it or don't want the data clutter.

idk, ready to be voted down. This definition was just not how I, personally, used it 😊

Reply:
I can understand the dev/test use case. For that I wrote a small mock APM server: https://gist.github.com/trentm/4b8c54a8bdb1c1eba2ade871253860f6   It shows all requests nicely, which can be helpful for dev as well:

```
[2021-06-29T23:18:13.255Z]  INFO: mockapmserver/47289 on pink.local: request (req.remoteAddress=::ffff:127.0.0.1, req.remotePort=58727, req.bodyLength=2067, res.bodyLength=2)
    POST /intake/v2/events HTTP/1.1
    accept: application/json
    user-agent: elasticapm-node/3.16.0 elastic-apm-http-client/9.8.1 node/12.22.1
    content-type: application/x-ndjson
    content-encoding: gzip
    host: localhost:8200
    connection: keep-alive
    transfer-encoding: chunked

    {
        "metadata": {
            "service": {
                "name": "play-s3",
                "environment": "development",
                ...
```

However, I think the user use case (the motivator for this is usage by Kibana core for https://github.com/elastic/kibana/issues/101711) either trumps the dev case, or we look into two separate config vars here.


@elastic/apm-agent-python What's your understanding/usage of `disable_send`?

Reply:
Sorry for the late reply. The reason we added `disable_send` back in the day was indeed to facilitate the use of the agent for local development and CI (not our CI, the user's CI). The other option is to disable the agent completely, which bears the risk that issues of the agent inter-playing with the app only show up on deployment.

Reply:
@beniwohli does that mean you're fine with the proposed semantics of `disable_send` or not? If not, how do you suggest moving forward? Introducing another config option that's specifically designed for the used cased mentioned here?

Reply:
@felixbarny sorry, I should have been more clear. I believe the original behavior (do everything but send data) still has merit and I think we should keep it.

I'm not sure we need to introduce another setting, as all the mentioned ways to limit overhead can already be achieved with existing settings AFAICT.

Reply:
That's two votes for _everything but sending_ 😅

Reply:
> as all the mentioned ways to limit overhead can already be achieved with existing settings AFAICT.

Not entirely, at least in the node.js agent. There isn't another way to avoid:
1. encoding of the span/transaction objects to data objects for sending
2. capturing an error (especially its stacktrace) for captured errors.

Reply:
Wouldn't `recording` or `enabled` do the trick? Maybe `source_lines_error_app_frames` for stacktraces?

Reply:
> Wouldn't `recording` or `enabled` do the trick?

The Node.js agent doesn't currently implement `recording` :) -- and I'm not entirely clear from the description of `recording` in the Python and Java agent docs that it matches. The [Java docs](https://www.elastic.co/guide/en/apm/agent/java/current/config-core.html#config-recording) imply `recording: false` means *not* tracking incoming HTTP requests, which means no trace-context propagation.

> Maybe `source_lines_error_app_frames` for stacktraces?

*Mostly*, yes. Stacktrace collection is the large part of error (and span) capture, but not all of it.

---

tl;dr: Would anyone object to the spec here saying something like "Agents **MAY** reduce runtime overhead where possible, for example skipping stack trace, metrics, error collection and breakdown metric calculation."? Effectively this would allow agent implementations to differ slightly without *yet* having get into specifying exactly how to support comparatively rare use cases. I imagine this particular Kibana use case -- using APM, but not for traces and metrics -- is rare. This is "option 1" in the details below.

Details (you can skip if your eyes are glazing over):

We discussed this at the apm-agents-nodejs call this morning and Alan and I chatted a bit about it. Here are options that I see:

1. Change the `disable_send` spec to say "agents **MAY** limit work ..." and basically have divergent use cases for the various agents.
2. Add a separate config option to more clearly state the use case for "do all your work, but be useful for CI where there isn't an APM server" (as Python and Ruby are using `disable_send=true` now). Seems unnecessary to have Python and Ruby teams prioritize this.
3. Add a separate config option to more clearly state the use case for "do the minimal work to support trace-context propagation and log correlation" (as Node.js and Java, sort of, are using `disable_send=true` now). I'm not sure what I would call this option.
4. Add separate config options to disable particular functionality (like span serialization and error capture) such that the "do the minimal work to support trace-context propagation and log correlation" use case is: disable_send=true, metrics_interval=0s, breakdown_metrics=false, serialize_objects_for_sending=false(?), capture_errors=false(?).

Reply:
1. This feels like under-specifying the option to a point where users have to rely on the implementation details of a specific agent.
2. I think the name `disable_send` conveys that quite nicely already.
3. To me, this sounds like a good option. Finding a better name that better suits the context propagation use case (`context_propagation_only`?) makes it easier to understand the intent of the option. Agents that have already implemented `disable_send` can just use it as an alias for the new config in the first phase. In a later stage, they can implement the listed optimizations.
The spec should be more concrete on the things agents SHOULD update (such as putting events in the queue, capturing errors and stack traces)
4. Seems like adding even more options and making it harder for the users to set the right set of options for the context propagation/log correlation use case.